module/os/linux/zfs/zfs_vfsops.c

```diff
@@ -273,39 +274,23 @@ zfs_sync(struct super_block *sb, int wait, cred_t *cr)
 	if (!wait)
 		return (0);
 
-	if (zfsvfs != NULL) {
-		/*
-		 * Sync a specific filesystem.
-		 */
-		dsl_pool_t *dp;
-		int error;
-
-		if ((error = zfs_enter(zfsvfs, FTAG)) != 0)
-			return (error);
-		dp = dmu_objset_pool(zfsvfs->z_os);
-
-		/*
-		 * If the system is shutting down, then skip any
-		 * filesystems which may exist on a suspended pool.
-		 */
-		if (spa_suspended(dp->dp_spa)) {
-			zfs_exit(zfsvfs, FTAG);
-			return (0);
-		}
-
-		if (zfsvfs->z_log != NULL)
-			zil_commit(zfsvfs->z_log, 0);
+	int err = zfs_enter(zfsvfs, FTAG);
+	if (err != 0)
+		return (err);
 
+	/*
+	 * If the pool is suspended, just return an error. This is to help
+	 * with shutting down with pools suspended, as we don't want to block
+	 * in that case.
+	 */
+	if (spa_suspended(zfsvfs->z_os->os_spa)) {
 		zfs_exit(zfsvfs, FTAG);
-	} else {
-		/*
-		 * Sync all ZFS filesystems.  This is what happens when you
-		 * run sync(1).  Unlike other filesystems, ZFS honors the
-		 * request by waiting for all pools to commit all dirty data.
-		 */
-		spa_sync_allpools();
+		return (SET_ERROR(EIO));
 	}
 
+	zil_commit(zfsvfs->z_log, 0);
```

Comment:
I notice this got rid of the check that z_log isn't NULL; is that always true?

Reply:
I am fairly sure, but only by code read. It's always set in `zfsvfs_setup()`, and NULLed in `zfsvfs_teardown()`. That can fail after NULLing it in very niche situations (something about unmounting during online recieve) but I'm not sure I see any way for a sync op to get here in that case, because they won't beat the teardown lock.

I'm not exactly brimming with confidence, as you see. If I leave the check in, then it seems we have to have a `txg_wait_synced()` call here instead to ensure anything was written out.